Content.Server/Atmos/Piping/Unary/EntitySystems/GasThermoMachineSystem.cs

```diff
@@ -58,26 +58,40 @@ private void OnThermoMachineUpdated(EntityUid uid, GasThermoMachineComponent the
             }
 
             float sign = Math.Sign(thermoMachine.Cp); // 1 if heater, -1 if freezer
-
-            // Implement hysteresis
-            float hystTemp = thermoMachine.TargetTemperature;
+            bool heater = sign > 0;
+            float targetTemp = thermoMachine.TargetTemperature;
+            float dTHyst = sign * thermoMachine.TemperatureTolerance;
+            float temp = inlet.Air.Temperature;
+
+            if (sign * temp >= sign * (targetTemp + dTHyst))
+                thermoMachine.HysteresisState = false;
+            if (sign * temp < sign * targetTemp)
+                thermoMachine.HysteresisState = true;
             if (thermoMachine.HysteresisState)
-            {
-                hystTemp += sign*thermoMachine.TemperatureTolerance;
-            }
+                targetTemp += dTHyst;
 
-            // Stop if we've reached our target temperature.
-            if (sign*inlet.Air.Temperature > sign*hystTemp) // inequality flips if freezer
+            float dT = targetTemp - temp;
+            if (sign != Math.Sign(dT))
             {
-                thermoMachine.HysteresisState = !thermoMachine.HysteresisState;
+                receiver.Load = 0f;
                 return;
             }
+            float Cin = _atmosphereSystem.GetHeatCapacity(inlet.Air);
 
             // Multiply power in by coefficient of performance, add that heat to gas
-            float dQ = receiver.Load * thermoMachine.Cp * args.dt;
-            _atmosphereSystem.AddHeat(inlet.Air, dQ);
+            float dQ = dT * Cin;
```

Comment:
Nitpick but this is the `dQLim`, the maximum dQ can be before it's clamped, swap the variable names.

Reply:
makes sense, done